### PR TITLE
Add Tags to Explorative Tracing Dashboard

### LINKED ISSUE #1971: Add tags to explorative tracing dashboard
- [x] Add tags to annotations
- [x] Clicking on tags lists them above the tracing list for reference / deselection and filters the list
- [x] Selecting multiple tags (could) filters them using an "AND" semantic
- [x] Persist selected tags across sessions in local storage
- [x] Consider using tracing type (skeleton/volume) and/or dataset name as (automatic) tags
- [x] Perhaps show last modified date instead of creation date
- [x] (Consider adding the search terms as temporary tags so that search terms persist)

<a href="https://timer.scm.io/repos/537237ce1a00001a003c4a14/issues/1971/create?referer=github" target="_blank">Log Time</a>

### PR DESCRIPTION
**Pull request by hotzenklotz**
### Mailable description of changes:
- Added a tagging system to tracings
- Tags can be easily searched and filtered for explorative tracing in the dashboard
- By default, the dataset name and "skeleton" / "volume" are added automatically as tags
- The "Archive All" button respects the current search / tag filtering now. Only the currently visible tracings are archived. A confirmation modal will display the number of affected tracings and ask for confirmation.

### Steps to test:
- Go to dashboard --> explorative tracings
- Add new custom tags by clicking on "plus icon".
- Reload page to see if they were saved
- Filter the list by clicking on several tags
- Reload page to see if the tag selection was persisted through localStorage
- Use the search bar to filter the list even further
- Use the search bar to add new "search tags". Type a search query and hit enter.
- Check that the list is filtered using an AND semantic

### Issues:
- fixes #1971 

@philippotto Please review the frontend changes.
@rschwanhold Please review the backend changes.

See  https://discuss.webknossos.org/t/improvements-to-the-explorative-annotations-dashboard-tab/408/14

------
- [x] Ready for review

### CODE REVIEW

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 1]
```diff
@@ -225,10 +241,11 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
   }
 
   archiveAll = () => {
+    const selectedAnnotations = this.getFilteredTracings();
     Modal.confirm({
-      content: "Are you sure you want to archive all explorative annotations?",
+      content: `Are you sure you want to archive all ${selectedAnnotations.length} explorative annotations matching the current search query / tags.?`,
       onOk: async () => {
-        const unarchivedAnnoationIds = this.state.unarchivedTracings.map(t => t.id);
+        const unarchivedAnnoationIds = selectedAnnotations.map(t => t.id);
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: Can you rename the variable so that it reflects that it holds the ids of the selected annotations? While doing this you can correct the `annoation` type :D

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 2]
```diff
@@ -241,21 +258,73 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
...
+  };
+
+  removeTagFromSearch = (tag: string): void => {
+    const newTags = this.state.tags.filter(t => t !== tag);
+    this.setState({ tags: newTags });
+    localStorage.setItem("lastDashboardSearchTags", JSON.stringify(newTags));
+  };
+
+  addTagToAnnotation = (annotation: APIAnnotationType, newTag: string): void => {
+    const newTracings = this.state.unarchivedTracings.map(t => {
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: You have to check here whether unarchived or archived tracings are currnetly shown. At the moment, archived tracings cannot receive a new tag.

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 3]
```diff
@@ -241,21 +258,73 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
...
+    this.setState({ unarchivedTracings: newTracings });
+  };
+
+  removeTagFromAnnotation = (
+    annotation: APIAnnotationType,
+    tag: string,
+    event: SyntheticInputEvent,
+  ): void => {
+    event.stopPropagation(); // prevent the onClick event
+    const newTracings = this.state.unarchivedTracings.map(t => {
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: Same here.

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 4]
```diff
@@ -241,21 +258,73 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
...
+      if (t.id === annotation.id) {
+        const newAnnotation = update(t, { tags: { $push: [newTag] } });
+
+        const url = `/annotations/${newAnnotation.typ}/${newAnnotation.id}/edit`;
+        Request.sendJSONReceiveJSON(url, { data: { tags: newAnnotation.tags } });
+        return newAnnotation;
+      }
+      return t;
+    });
+    this.setState({ unarchivedTracings: newTracings });
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: Same here.

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 5]
```diff
@@ -241,21 +258,73 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
...
+  };
+
+  removeTagFromAnnotation = (
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: Can you refactor the addTag and removeTag functions so that they use a single generic implementation? The only difference between the two is the operation on oldTags + the given tag (either `without` or `push`).

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 6]
```diff
@@ -321,12 +406,28 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
...
+
   render() {
     const marginRight = { marginRight: 8 };
     const search = (
       <Search
         style={{ width: 200, float: "right" }}
-        onPressEnter={this.handleSearch}
+        onPressEnter={(event: SyntheticInputEvent) => {
+          this.addTagToSearch(event.target.value);
+          this.setState({ searchQuery: "" });
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: This should clear the search box too, which is why `searchQuery` should be passed to `<Search />`.

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js [hunk 7]
```diff
@@ -360,6 +461,7 @@ export default class ExplorativeAnnotationsView extends React.PureComponent {
               {search}
             </div>}
         <h3>Explorative Annotations</h3>
+        {this.renderSearchTags()}
```
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: Do you think it's clear to everyone how the tag search works? Maybe we should add a label like "Current tag filters" or a tooltip?

File: app/assets/javascripts/oxalis/store.js [hunk 8]
```diff
@@ -201,6 +203,7 @@ export type ReadOnlyTracingType = {
   +boundingBox: ?BoundingBoxType,
   +restrictions: RestrictionsType & SettingsType,
   +isPublic: boolean,
+  +tags: Array<string>,
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/store.js]: We should really refactor these types. BoundingBox, restrictions, isPublic, tags (and probably more) seem to exist in all tracing types.

File: app/assets/javascripts/oxalis/view/components/editable_text_icon.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    if (this.state.isEditing) {
+      return (
+        <Input
+          value={this.state.value}
+          onChange={this.handleInputChange}
+          onPressEnter={this.handleOnChange}
+          onBlur={this.handleOnChange}
+          style={{ width: 75 }}
+          size="small"
+          autoFocus
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/view/components/editable_text_icon.js]: Nice.

File: conf/evolutions/36.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+// Update to support tags in annotations
```
Comment (philippotto) [inline conf/evolutions/36.js]: Why is this file here twice (35 and 36)?

File: app/assets/javascripts/oxalis/view/components/editable_text_icon.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  render() {
+    const iconStyle = { cursor: "pointer" };
+
+    if (this.state.isEditing) {
+      return (
+        <Input
+          value={this.state.value}
+          onChange={this.handleInputChange}
+          onPressEnter={this.handleOnChange}
+          onBlur={this.handleOnChange}
```
Comment (rschwanhold) [inline app/assets/javascripts/oxalis/view/components/editable_text_icon.js]: It might be a bit confusing that `onPressEnter` and `onBlur `are `{this.handleOnChange}`, while `onChange` is not `{this.handleOnChange}` but `{this.handleInputChange}`. My point is that the name `onChange` might be missleading

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js (hunk 1 above)
Comment (hotzenklotz) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: It does not hold the ids. It holds the fully annotation objects, though.

File: app/assets/javascripts/oxalis/view/components/editable_text_icon.js (hunk 9 above)
Comment (hotzenklotz) [inline app/assets/javascripts/oxalis/view/components/editable_text_icon.js]: I just learned that yesterday. I had used it previously but forgot about this. A lot easier than with React refs.

File: conf/evolutions/36.js (hunk 10 above)
Comment (hotzenklotz) [inline conf/evolutions/36.js]: My bad.

File: app/assets/javascripts/oxalis/view/components/editable_text_icon.js (hunk 11 above)
Comment (hotzenklotz) [inline app/assets/javascripts/oxalis/view/components/editable_text_icon.js]: Fair point. I will change it.

File: app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js (hunk 1 above)
Comment (philippotto) [inline app/assets/javascripts/dashboard/views/dash/explorative_annotations_view.js]: I meant the `unarchivedAnnoationIds` variable which holds the ids of the selected annotations objects and should be named `selectedAnnotationIds` or could be directly inlined where it's used since the extraction into an own variable doesn't add much value here (it's only used once anyway).
